### PR TITLE
Browser.waitForTarget

### LINKED ISSUE #739: Browser.waitForTarget
https://github.com/kblok/puppeteer/pull/13/commits/6ac66c3547fdd95fed97fc90b15ac0b4bf178a86

### PR DESCRIPTION
**Pull request by bdurrani**
closes #739

### CODE REVIEW

**Comment by kblok** [comment]
@bdurrani though it doesn't affect the final result. Try to start every new feature branching from master, so we don't get 17 commits when we should've got only one.

Take a look at the CodeFactor error.

File: lib/PuppeteerSharp/Browser.cs [hunk 1]
```diff
@@ -258,6 +258,62 @@ public async Task<string> GetUserAgentAsync()
         /// <returns>Task</returns>
         public Task CloseAsync() => _closeTask ?? (_closeTask = CloseCoreAsync());
 
+        /// <summary>
+        ///
```
Comment (kblok) [inline lib/PuppeteerSharp/Browser.cs]: Missing comment

File: lib/PuppeteerSharp/Browser.cs [hunk 2]
```diff
@@ -258,6 +258,62 @@ public async Task<string> GetUserAgentAsync()
         /// <returns>Task</returns>
         public Task CloseAsync() => _closeTask ?? (_closeTask = CloseCoreAsync());
 
+        /// <summary>
+        /// 
+        /// </summary>
+        public int DefaultWaitForTimeout { get; set; } = 30000;
+
+        /// <summary>
+        ///
```
Comment (kblok) [inline lib/PuppeteerSharp/Browser.cs]: Missing comments

File: lib/PuppeteerSharp/Browser.cs [hunk 3]
```diff
@@ -258,6 +258,62 @@ public async Task<string> GetUserAgentAsync()
...
+                if (predicate(e.Target))
+                {
+                    targetCompletionSource.TrySetResult(e.Target);
+                }
+            }
+
+            try
+            {
+                TargetCreated += TargetCreatedHandler;
+                TargetChanged += TargetChangedHandler;
```
Comment (kblok) [inline lib/PuppeteerSharp/Browser.cs]: Do we need two handlers?

File: lib/PuppeteerSharp/Browser.cs [hunk 4]
```diff
@@ -258,6 +258,62 @@ public async Task<string> GetUserAgentAsync()
...
+                await Task.WhenAny(new[]
+                {
+                    TaskHelper.CreateTimeoutTask(timeout),
+                    targetCompletionSource.Task
+                }).ConfigureAwait(false);
+
```
Comment (kblok) [inline lib/PuppeteerSharp/Browser.cs]: Here we need to await the winning task so we fail if it's the timeout.

```cs 
var task = await Task.WhenAny(new[]
{
    TaskHelper.CreateTimeoutTask(timeout),
    targetCompletionSource.Task
 }).ConfigureAwait(false);
await task;
```

File: lib/PuppeteerSharp/BrowserContext.cs [hunk 5]
```diff
@@ -56,6 +56,23 @@ internal BrowserContext(IConnection connection, Browser browser, string contextI
         /// <returns>An array of all active targets inside the browser context</returns>
         public Target[] Targets() => Array.FindAll(Browser.Targets(), target => target.BrowserContext == this);
 
+
+        /// <summary>
+        /// This searches for a target in this specific browser context.
+        /// <example>
+        /// <code>
+        /// <![CDATA[
+        /// TODO
```
Comment (kblok) [inline lib/PuppeteerSharp/BrowserContext.cs]: Missing

File: lib/PuppeteerSharp/BrowserContext.cs [hunk 6]
```diff
@@ -56,6 +56,23 @@ internal BrowserContext(IConnection connection, Browser browser, string contextI
...
+        /// ]]>
+        /// </code>
+        /// </example>
+        /// </summary>
+        /// <param name="predicate">A function to be run for every target</param>
+        /// <param name="options">options</param>
+        /// <returns></returns>
+        public async Task<Target> WaitForTargetAsync(Func<Target, bool> predicate, WaitForOptions options = null) 
+            => await Browser.WaitForTargetAsync((target) => target.BrowserContext == this && predicate(target), options);
```
Comment (kblok) [inline lib/PuppeteerSharp/BrowserContext.cs]: We can just return the task

**Comment by kblok** [comment]
@bdurrani you can do something like this
```
git checkout upstream/master -b some-feature --no-track
```

**Comment by bdurrani** [comment]
> @bdurrani you can do something like this
> 
> ```
> git checkout upstream/master -b some-feature --no-track
> ```

I'll give this a shot! I'm aiming to wrap up this PR by the end of the day. Thanks

File: lib/PuppeteerSharp.Tests/BrowserContextTests/BrowserContextTests.cs [hunk 7]
```diff
@@ -149,5 +149,28 @@ public async Task ShouldWorkAcrossSessions()
...
+            var targetPage = await promiseTarget.PageAsync();
+            Assert.Equal(targetPage, page);
+            await context.CloseAsync();
+        }
+
+        [Fact]
+        public async Task ShouldTimeoutWaitingForNonExistantTarget()
+        {
+            var context = await Browser.CreateIncognitoBrowserContextAsync();
+            var targetPromise = context.WaitForTargetAsync((target) => target.Url == TestConstants.EmptyPage);
```
Comment (kblok) [inline lib/PuppeteerSharp.Tests/BrowserContextTests/BrowserContextTests.cs]: We don't need `targetPromise`

**Comment by bdurrani** [comment]
strange.. That failing test was passing previously and the only change i made was remove that one redundant line in the test..

**Comment by kblok** [comment]
@bdurrani it was a hiccup :)

**Pull request merged**